src/api/MetadataQuery.js

```diff
@@ -147,54 +158,84 @@ class MetadataQuery extends Base {
      * @param {Object} - metadata query response object
      * @return {Object} - flattened metadata query response object
      */
-    flattenMetdataQueryResponse = ({ entries, next_marker }: MetadataQueryResponse): FlattenedMetadataQueryResponse => {
+    flattenMetdataQueryResponse = (): FlattenedMetadataQueryResponse => {
+        const { entries, next_marker }: MetadataQueryResponse = this.metadataQueryResponse;
         return {
             items: entries.map<FlattenedMetadataQueryResponseEntry>(this.flattenResponseEntry),
             nextMarker: next_marker,
         };
     };
 
     /**
-     * @param {Object} response
+     * Success handler function storing flattened response in Cache
+     * @param {Object} templateSchemaResponse - metadata template schema response
      */
-    queryMetadataSuccessHandler = ({ data }: { data: MetadataQueryResponse }): void => {
+    successHandler = (templateSchemaResponse: ?MetadataTemplateSchemaResponse): void => {
+        this.metadataTemplate = getProp(templateSchemaResponse, 'data');
         const cache: APICache = this.getCache();
-        const filteredResponse = this.filterMetdataQueryResponse(data);
         // Flatten the filtered metadata query response and set it in cache
-        cache.set(this.key, this.flattenMetdataQueryResponse(filteredResponse));
+        cache.set(this.key, this.flattenMetdataQueryResponse());
         this.finish();
     };
 
+    /**
+     * Gets template schema
+     * @param {Object} response
+     */
+    getTemplateSchemaInfo = ({
+        data,
+    }: {
+        data: MetadataQueryResponse,
+    }): Promise<MetadataTemplateSchemaResponse | void> => {
+        this.metadataQueryResponse = this.filterMetdataQueryResponse(data);
+        const { entries } = data;
+
+        if (!entries || entries.length === 0) {
+            // if no items found for provided query, don't make API call to get template info
+            return Promise.resolve();
+        }
+
+        const metadata = getProp(entries, '[0].metadata');
+        this.templateScope = Object.keys(metadata)[0];
+        const instance = metadata[this.templateScope];
+        this.templateKey = Object.keys(instance)[0];
+
+        return this.metadataAPI.getSchemaByTemplateKey(this.templateKey);
+    };
+
     /**
      * Does the network request to metadata query API
      * @param {Object} query query object with SQL Clauses like properties
      * @return {void}
      */
-    queryMetadataRequest(query: MetadataQueryType): void {
+    queryMetadataRequest(query: MetadataQueryType): Promise<void> {
         if (this.isDestroyed()) {
-            return;
+            Promise.resolve();
```

Comment:
What purpose is this serving if this Promise isn't returned?

Reply:
Doesn't matter. Undoing this change.

Reply:
I see. Will the calling code need to check for the existence of the promise before adding to its chain?